src/modules/roc_sdp/session_description.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_sdp/session_description.h"
+
+namespace roc {
+namespace sdp {
+
+SessionDescription::SessionDescription(core::IAllocator& allocator)
+    : guid_(allocator)
+    , media_descriptions_()
+    , allocator_(allocator) {
+    clear();
+}
+
+void SessionDescription::clear() {
+    guid_.clear();
+    origin_unicast_address_.clear();
+    session_connection_data_.clear();
+
+    while (media_descriptions_.size() > 0) {
+        core::SharedPtr<MediaDescription> m = media_descriptions_.back();
+        media_descriptions_.remove(*m);
+    }
+}
+
+const char* SessionDescription::guid() const {
+    if (guid_.is_empty()) {
+        return NULL;
+    }
+    return guid_.c_str();
+}
+
+bool SessionDescription::set_guid(const char* start_p_origin_username,
+                                  const char* end_p_origin_sess_id,
+                                  const char* start_p_origin_nettype,
+                                  const char* end_p_origin_addr) {
```

Comment:
From the RFC:
"In general, the "o=" field serves as a globally unique identifier for this version of this session description, and the subfields excepting the version taken together identify the session irrespective of any modifications". 

So I remove the  `sess-version` part of the origin field, that is between the `username` and the `sess-id`, to create the GUID.

But I can add it, and the GUID will be just start_p_origin_username and end_p_origin_addr (so the entire origin field).

What do you think ?

From the LogInfo i put in this function, addrtype seems to be taken into account in the GUID.

Reply:
Excluding `sess-version` from GUID is correct.

What I'm asking, why do we, for example, add the entire string `<username> <sess-id>` (`append_str_range(start_p_origin_username, end_p_origin_sess_id)`) instead of adding `<username>` and `<sess-id>` separately, excluding separator (space) between them?

The current approach assumes that the separator is always the same, e.g. is always exactly one space, and it can not be two spaces, or other white-space character, like tabulation. Is this true and required by RFC? If yes, the code is good, otherwise, two logically equal descriptions, which differ only by some delimiters, may produce different GUIDs, which is bad.

Reply:
> From the LogInfo i put in this function, addrtype seems to be taken into account in the GUID.

Indeed, my bad.

Reply:
Yeah, and even if RFC requires exactly one space, does it make sense to be ready for multiple spaces as delimiters anyway? I can imagine that some implementation may produce extra spaces, though I'm not aware of such. Usually it's a good idea to be generous when parsing and allow "almost" correct input while it remains unambiguous. Would such behavior break other things?

Reply:
Yeah you're right, it's better to be ready for multiple spaces or a tab as delimiters.

Reply:
Added this to https://github.com/roc-project/roc/issues/200#issuecomment-612383960